ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs

```diff
@@ -1217,20 +1221,39 @@ run tracers tracersExtra args argsExtra apps appsExtra = do
 -- Utility Function
 --
 
-computePeerSharingPeers :: (MonadSTM m)
+
+-- | Select a random subset of the known peers that are available to publish through peersharing.
+-- The list of peers will change after `policyPeerShareRetryTime` seconds.
+-- The list of peers shared does at most change by the number of peers added or removed.
+-- That is a newly added or removed peer can at most lead to one new peer beeing shared.
+--
+computePeerSharingPeers :: (  MonadSTM m
+                            , MonadMonotonicTime m
+                            , Hashable ntnAddr
+                           )
                         => STM m (PublicPeerSelectionState ntnAddr)
                         -> StrictTVar m StdGen
+                        -> StrictTVar m Time
+                        -> PeerSelectionPolicy ntnAddr m
                         -> PeerSharingAmount
                         -> m [ntnAddr]
-computePeerSharingPeers readPublicState genVar amount = do
+computePeerSharingPeers readPublicState genVar reSaltAtVar PeerSelectionPolicy{..} amount = do
+  now <- getMonotonicTime
   publicState <- atomically readPublicState
-  gen <- atomically $ stateTVar genVar split
+  salt <- atomically $ do
+    reSaltAt <- readTVar reSaltAtVar
+    if reSaltAt <= now
+       then do
+         writeTVar reSaltAtVar $ addTime policyPeerShareRetryTime now
```

Comment:
The use of `policyPeerShareRetryTime ` interacts poorly with how we use it on the gosping side. Basically nodes will request new peers too early because the retry time is set when we send a request, not when the node gets a reply.
Will switch to a different timeout that is less than `policyPeerShareRetryTime ` .

Reply:
Could you create an issue for it.

Reply:
No, I prefer to fix it in this PR.